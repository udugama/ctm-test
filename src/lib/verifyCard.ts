import * as Ajv from 'ajv';
import * as requestSchema from './../schema/requestCardValidate.json'
import * as responseSchema from './../schema/responseCardValidate.json'
import { getCreditCardModel } from './models/Cards';
import { checkCardNumberLength, validateCardNumber } from './validations';

export enum CardType {
    Visa = 'Visa',
    MasterCard = 'MasterCard',
    Discover = 'Discover',
    AMEX = 'AMEX',
};
export enum IDocument { creditcard };
export enum Status {
    valid = 'valid',
    invalid = 'invalid',
}

export declare interface IPayload {
    customerId: number;
    document: IDocument;
    name: string;
    type: CardType;
    number: number;
    csv: number;
}

export declare interface IResponse {
    status: Status;
    content: IPayload;
}

/**
 * saves pet to the database and return database response
 * @param body
 */
export const saveCreditCard = (async (payload: IPayload) => {
    const CreditCardModel: any = await getCreditCardModel();

    const CreditCard: any = new CreditCardModel(payload);

    const creditCard: any = await CreditCard.save()
    .catch((error) => {
      console.log('Error saving pet.', error);
      throw new Error(error);
    });

    return creditCard;
});

const validate = (body: object, schema: any, ctx: any, name: string) => {
    // validate the request agains the json schema to ensure intergrity of request payload 
    const ajv: any = new Ajv ({ allErrors: true });
    const valid: boolean = ajv.validate(schema, body); 
    if (!valid) {
        console.log(`${name} schema validation errors : `, ajv.errors);
        ctx.throw(422, `${name} validation error(s): ` + JSON.stringify(ajv.errors));
    }
}

/**
 * validate the request agains the create schema and process create response 
 * @param ctx 
 */
export const verifyCreditCard = (async (ctx)=> {
    let status: Status = Status.invalid; 
    const { request: { body, header } } = ctx;

    const requestPayload: IPayload = body;

    // check content type header
    if (header['content-type'] !== 'application/json') {
        ctx.throw(422, 'Incorrect `content-type` header recieved, expect `application/json`');
    }

    console.log('request payload : ', requestPayload);

    // validate request agains request schema
    validate(requestPayload, requestSchema, ctx, 'Request');

    // Validate credit Card numbers
    const validateCardType = checkCardNumberLength(requestPayload.type, requestPayload.number);
    const validCardNumber = validateCardNumber(requestPayload.number);

    if ( validateCardType || validCardNumber ) {
        status = Status.valid;
    }

    const { customerId, document, type, name, number, csv } = await saveCreditCard(requestPayload);

    const content = <IPayload> { customerId, document, type, name, number, csv };

    const respsonse = <IResponse> {
        status,
        content: content,
    };

    // validate request agains request schema
    validate(respsonse, responseSchema, ctx, 'Response');

    ctx.body = respsonse;
});
