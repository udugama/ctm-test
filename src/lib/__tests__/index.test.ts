import * as clone from 'lodash.clone';
import { saveCreditCard, verifyCreditCard, Status } from '../verifyCard';
import { saveCardResponse, body } from './fixtures/requests';
import { server } from '../../index';
import * as request from 'supertest';

describe('POST creditcard/verify', () => {
    afterAll(() => {
        server.close();
    });
    it('is a function', () => {
        expect(verifyCreditCard).toBeInstanceOf(Function);
    });

    test('successfully response', async () => {
        (saveCreditCard as any) = jest
            .fn()
            .mockReturnValue(saveCardResponse);
        await (request.agent(server) as any)
        .post('/creditcard/verify')
        .send(body)
        .set('Content-Type', 'application/json')
        .expect('Content-Type', 'application/json; charset=utf-8')
        .expect(200)
        .then((response) => {
            expect(response.body.status).toEqual(Status.valid);
            expect(response.body.status).toMatchSnapshot();
        });
    });

    test('successful response with invalid status - invalid type validation', async () => {
        const invalidBody: any = clone(body);
        invalidBody.number = 5408041234567893; // MasterCard type of number as a Visa
        (saveCreditCard as any) = jest
            .fn()
            .mockReturnValue(saveCardResponse);
        await (request(server) as any)
        .post('/creditcard/verify')
        .send(invalidBody)
        .set('Content-Type', 'application/json')
        .expect('Content-Type', 'application/json; charset=utf-8')
        .expect(200)
        .then((response) => {
            expect(response.body.status).toEqual(Status.invalid);
            expect(response.body).toMatchSnapshot();
        }) 
    });

    test('successful response with invalid status - invalid number', async () => {
        const invalidBody: any = clone(body);
        invalidBody.number = 408941234567893; // incorect Visa number
        (saveCreditCard as any) = jest
            .fn()
            .mockReturnValue(saveCardResponse);
        await (request(server) as any)
        .post('/creditcard/verify')
        .send(invalidBody)
        .set('Content-Type', 'application/json')
        .expect('Content-Type', 'application/json; charset=utf-8')
        .expect(200)
        .then((response) => {
            expect(response.body.status).toEqual(Status.invalid);
            expect(response.body).toMatchSnapshot();
        });
    });

    test('error response with invalid type - schema error - status 422', async () => {
        const invalidBody: any = clone(body);
        invalidBody.type = 'Veeesa'; // incorect card type
        (saveCreditCard as any) = jest
            .fn()
            .mockReturnValue(saveCardResponse);
        await (request(server) as any)
        .post('/creditcard/verify')
        .send(invalidBody)
        .set('Content-Type', 'application/json')
        .expect(422)
    });

    test('error response aditional field - schema error - status 422', async () => {
        const invalidBody: any = clone(body);
        invalidBody.aaaaa = 'abcd'; // aditional field inn the request
        (saveCreditCard as any) = jest
            .fn()
            .mockReturnValue(saveCardResponse);
        await (request(server) as any)
        .post('/creditcard/verify')
        .send(invalidBody)
        .set('Content-Type', 'application/json')
        .expect(422)
    });

    test('response schema validation - schema error - status 422', async () => {
        const invalidBody: any = clone(body);
        invalidBody.aaaaa = 'abcd'; // aditional field inn the request
        (saveCardResponse as any).type = 'xyz';
        (saveCreditCard as any) = jest
            .fn()
            .mockReturnValue(saveCardResponse);
        await (request(server) as any)
        .post('/creditcard/verify')
        .send(invalidBody)
        .set('Content-Type', 'application/json')
        .expect(422)
    });
});