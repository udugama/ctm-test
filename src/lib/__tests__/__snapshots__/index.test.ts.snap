// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`POST creditcard/verify successful response with invalid status - invalid number 1`] = `
Object {
  "content": Object {
    "csv": 120,
    "customerId": 1,
    "document": "creditcard",
    "name": "Prasad S Udugama",
    "number": 4408041234567893,
    "type": "Visa",
  },
  "status": "invalid",
}
`;

exports[`POST creditcard/verify successful response with invalid status - invalid type validation 1`] = `
Object {
  "content": Object {
    "csv": 120,
    "customerId": 1,
    "document": "creditcard",
    "name": "Prasad S Udugama",
    "number": 4408041234567893,
    "type": "Visa",
  },
  "status": "invalid",
}
`;

exports[`POST creditcard/verify successfully response 1`] = `"valid"`;
